6462 Update front end pugins interface

andreievg opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

This PR is ready for review but I ~~haven't self reviewed it yet sorry~~ or fixed unit tests (fixed downstream), downstream PR has all of the changes in the stacked PRs, with documentation at some test examples: https://github.com/msupply-foundation/open-msupply/pull/6485

Fixes #6462

# 👩🏻‍💻 What does this PR do?

Refactor plugin interface to match examples here: https://github.com/andreievg/plugins-example?tab=readme-ov-file

* Use object based plugin interface
* Allows for a plugin to implement multiple areas, multiple times
* Single plugin interface can require multiple components/functions/constants (like the column which also gets StateLoader to load all data in one go rather then by row)
* Event replaced with simpler hook that is passed to the plugin component to (simpler way to update isDirty state, and do an action onSave, which can be awaited with direct feedback to the caller on the core end)
* Not using lazy component, fetch all plugins at the start of the app, they will become available when loaded

andreievg commented on client/packages/common/src/plugins/api/operations.generated.ts:
```diff
@@ -0,0 +1,33 @@
+
+export const FrontendPluginMetadataDocument = gql`
+    query frontendPluginMetadata {
+  frontendPluginMetadata {
```
Previous was:
```
query plugins {
  plugins {
    config
    name
    path
  }
```

Updated to "frontendPluginMetadata", sorry extra change I probably could have avoided

andreievg commented on client/packages/common/src/plugins/pluginProvider.ts:
```diff
@@ -0,0 +1,113 @@
+
+// PLUGINS INIT
+
+// LOAD REMOTE PLUGIN
```
This is functionality similar to what was there before, most of the code is taking from federation module examples, and based on: https://github.com/andreievg/plugins-example

There is no lazy loading anymore, which simplifies things a bit

andreievg commented on client/packages/common/src/plugins/types.ts:
```diff
@@ -1,68 +1,23 @@
-
-export type PluginModule<DataType> = {
-  default: React.ComponentType<{ data?: DataType | undefined }>;
+export type Plugins = {
```
Simplified types and easier to add new interface, and it's implementation in plugin

andreievg commented on client/packages/common/src/plugins/types.ts:
```diff
@@ -1,68 +1,23 @@
-export type PluginDefinition = {
-  columnPlugins: ColumnPlugin[];
-  componentPlugins: ComponentPlugin[];
+type PluginData<D> = { relatedRecordId: string; data: D };
```
This is used in columns plugin, as a type helper for plugin store (thought it would be shared in other areas)

andreievg commented on client/packages/common/src/plugins/usePluginEvents.ts:
```diff
@@ -0,0 +1,40 @@
+  input: EventInput
+) => Promise<EventResult>;
+
+export const usePluginEvents = <EventInput, EventResult>() => {
```
Events are changed to be a prop of this hook, please see documentation, also upstream `isDirty` state is change to be a more generic 'state'

andreievg commented on client/packages/common/src/utils/types/TypeUtils.ts:
```diff
@@ -21,3 +21,5 @@ export const getErrorMessage = (error: unknown): string => {
   return String(error);
 };
+
+export type ArrayElement<T> = T extends (infer U)[] ? U : T;
```
Quite useful helper to extract type from array

andreievg commented on client/packages/dashboard/src/DashboardService.tsx:
```diff
@@ -24,7 +22,7 @@ const Dashboard: React.FC = () => {
       <DistributionWidget />
       <StockWidget />
-      {plugins}
+      {plugins.dashboard?.map((Plugin, index) => <Plugin key={index} />)}
```
This is the new 'binding' for plugins to core areas

andreievg commented on client/packages/host/getLocalPlugins.js:
```diff
@@ -0,0 +1,43 @@
+// Script expects the file structure to be:
+// packages/plugins/{repository or plugin name}/frontend/latest
+
+const fs = require('fs');
```
The reason for this file, is mainly for versioning up the track, when wanted to load an older version of the plugin to test with older version of omSup

andreievg commented on client/packages/host/src/api/api.ts:
```diff
@@ -17,12 +17,6 @@ export const getHostQueries = (sdk: Sdk) => ({
       const result = await sdk.databaseSettings();
       return result.databaseSettings;
     },
-    plugins: async () => {
```
this was relocated to common

andreievg commented on client/packages/host/src/useInitPlugins.ts:
```diff
@@ -0,0 +1,48 @@
+import {
```
I wanted to keep this file in common but had to be moved to host, small comment below, something about loading local plugins with dynamic import was being tricky when dynamic import is not in the package that is being built/run

andreievg commented on client/packages/host/webpack.config.js:
```diff
@@ -174,6 +142,13 @@ module.exports = env => {
         name: 'host',
         shared: [
           {
+            '@openmsupply-client/common': {
```
This one had to be shared, otherwise context are not filled/shared (even though react-singleton-context is singleton and shared)

andreievg commented on client/packages/plugins/plugin-webpack.config.js:
```diff
@@ -0,0 +1,73 @@
+
+const prod = process.env.NODE_ENV === 'production';
+
+module.exports = ({ pluginName, distDir }) => ({
```
Webpack config is shared for plugins, this way plugins stay a bit 'lighter'

andreievg commented on client/packages/plugins/plugin-webpack.config.js:
```diff
@@ -0,0 +1,73 @@
+  devtool: prod ? undefined : 'source-map',
+  plugins: [
+    new ModuleFederationPlugin({
+      name: pluginName,
```
Using pluginName/code coming from plugins package.json

andreievg commented on client/packages/system/src/Stock/ListView/ListView.tsx:
```diff
@@ -160,13 +158,16 @@ const StockListComponent: FC = () => {
     <>
       <Toolbar filter={filter} />
       <AppBarButtons />
+      {plugins.stockColumn?.StateLoader?.map((StateLoader, index) => (
```
There is a state loader for columns, which allows to get all data for all columns in one go

lache-melvin reviewed (approved):
LGTM, looks like a few things have changed upstream 🙏

jmbrunskill commented on client/packages/plugins/examples:
Yes, I've done this on my branch and it's working fine

andreievg commented:
@lache-melvin thank you for review !

> LGTM, looks like a few things have changed upstream

I am sorry, I was trying to not change too much in follow up stacked PRs

@jmbrunskill, added gitignore

32bb3b1b8554ede5e9cf75d06daf51a0803a74a5

> Might be good to mention in docs somewhere that order: blah in the column definition is why this can render not last...

Added TODO for now, thanks: 2292e106ba76b4009fb31c5df7d8f945216625a7

andreievg commented:
Closing no new commits that are not in #6485
